types/src/transaction/authenticator.rs

```diff
@@ -820,11 +831,23 @@ pub struct MultiKey {
 }
 
 impl MultiKey {
-    pub fn new(public_keys: Vec<AnyPublicKey>, signatures_required: u8) -> Self {
-        Self {
+    pub fn new(public_keys: Vec<AnyPublicKey>, signatures_required: u8) -> Result<Self> {
+        ensure!(
+            signatures_required > 0,
+            "The number of required signatures is 0."
+        );
+
+        ensure!(
```

Comment:
if we want to enforce this, it's not sufficient in constructor. we directly deserialize the multikey from network, it needs to be checked either in a verify function or deserialization.

although I think it might be impossible to enforce it, move only takes a hash of authenticator for key rotations, we don't have access to the actual authenticator value

Reply:
Hm...

It is indeed impossible to enforce, since anyone can send a payment to a non-existing **invalid** multiauth address (e.g., with 0 signatures required).

This is not unique to multiauth addresses though.

Even, say, Ed25519 addresses have the same problem: anyone can send a payment to a non-existing **invalid** Ed25519 address (e.g., the SK associated with the address's underlying PK is unknown).

It will not be possible to check anything about `MultiKey` during deserialization, since that happens behind our backs via `serde`...

So you are right: we need to duplicate some of these checks in `MultiKeyAuthenticator::verify`...

Reply:
cc @davidiw 

I would add the following to `MultiKeyAuthenticator::verify`:

```
diff --git a/types/src/transaction/authenticator.rs b/types/src/transaction/authenticator.rs
index 492a5c179f..379adf80cb 100644
--- a/types/src/transaction/authenticator.rs
+++ b/types/src/transaction/authenticator.rs
@@ -785,11 +785,20 @@ impl MultiKeyAuthenticator {
     }

     pub fn verify<T: Serialize + CryptoHash>(&self, message: &T) -> Result<()> {
+        ensure!(
+            self.public_keys.signatures_required() > 0,
+            "The number of required signatures is 0."
+        );
+        ensure!(
+            self.public_keys.len() >= self.public_keys.signatures_required() as usize,
+            "The number of public keys is smaller than the number of required signatures, {} < {}",
+            self.public_keys.len(),
+            signatures_required
+        );
         ensure!(
             self.signatures_bitmap.last_set_bit().is_some(),
             "There were no signatures set in the bitmap."
         );
-
         ensure!(
             (self.signatures_bitmap.last_set_bit().unwrap() as usize) < self.public_keys.len(),
             "Mismatch in the position of the last signature and the number of PKs, {} >= {}.",
```

This would prevent accessing [accidentally-created?] multiauth accounts with either (1) 0 required signatures or (2) less PKs than required signatures.

For (2), the other logic in `MultiKeyAuthenticator::verify` would prevent access anyway. 

But for (1) is there a reason we might want to allow _anyone_ to spend/access a 0-out-of-n multiauth account?